File: exercises/ddd_infrastructure_layer/java/src/main/java/application/services/SubmitMotivationLetterService.java
package application.services;

import java.util.UUID;

import com.google.inject.Inject;

import domain.application.Application;
import domain.application.ApplicationId;
import domain.application.ApplicationRepository;
import domain.application.MotivationLetter;

public class SubmitMotivationLetterService {
    private ApplicationRepository repo;

    @Inject
    public SubmitMotivationLetterService(ApplicationRepository repo) {
        this.repo = repo;
    }

    public void execute(SubmitMotivationLetterCommand cmd) throws Exception {
        ApplicationId id = new ApplicationId(cmd.applicationId());
        Application app = repo.getById(id);
        if (app == null) {
            throw new RuntimeException("Application not found");
        }

        try {
            MotivationLetter motivationLetter = new MotivationLetter(cmd.motivationLetter());
            app.fillOutMotivationLetter(motivationLetter);
        } catch (Exception e) {
            throw new RuntimeException("Failed to fill out motivation letter: " + e.getMessage());
        }

        repo.save(app);

    }

    public static record SubmitMotivationLetterCommand(UUID applicationId, String motivationLetter) { }
}

File: exercises/ddd_infrastructure_layer/typescript/src/application/services/SubmitMotivationLetterService.ts
import type { ApplicationRepository } from '../../domain/application/ApplicationRepository';
import { ApplicationId } from '../../domain/application/ApplicationId';
import { MotivationLetter } from '../../domain/application/MotivationLetter';

export interface SubmitMotivationLetterCommand {
  applicationId: string; // UUIDv4
  motivationLetter: string;
}

export class SubmitMotivationLetterService {
  private repo: ApplicationRepository;

  constructor(repo: ApplicationRepository) {
    this.repo = repo;
  }

  async execute(cmd: SubmitMotivationLetterCommand): Promise<void> {
    const id = new ApplicationId(cmd.applicationId);
    const app = await this.repo.getById(id);
    if (!app) {
      throw new Error('Application not found');
    }

    try {
      const motivationLetter = new MotivationLetter(cmd.motivationLetter);
      app.submitMotivationLetter(motivationLetter);
    } catch (e) {
      throw new Error(`Failed to fill out motivation letter: ${(e as Error).message}`);
    }

    await this.repo.save(app);
  }
}
